README.rst
Eclipse Paho™ MQTT Python Client
================================

This document describes the source code for the `Eclipse Paho <http://eclipse.org/paho/>`_ MQTT Python client library, which implements versions 3.1 and 3.1.1 of the MQTT protocol.

This code provides a client class which enable applications to connect to an `MQTT <http://mqtt.org/>`_ broker to publish messages, and to subscribe to topics and receive published messages. It also provides some helper functions to make publishing one off messages to an MQTT server very straightforward.

It supports Python 2.7 or 3.2+, with limited support for Python 2.6.

The MQTT protocol is a machine-to-machine (M2M)/"Internet of Things" connectivity protocol. Designed as an extremely lightweight publish/subscribe messaging transport, it is useful for connections with remote locations where a small code footprint is required and/or network bandwidth is at a premium.

Paho is an `Eclipse Foundation <https://www.eclipse.org/org/foundation/>`_ project.


Contents
--------

* Installation_
* `Usage and API`_
    * `Client`_
        * `Constructor / reinitialise`_
        * `Option functions`_
        * `Connect / reconnect / disconnect`_
        * `Network loop`_
        * `Publishing`_
        * `Subscribe / Unsubscribe`_
        * `Callbacks`_
        * `External event loop support`_
        * `Global helper functions`_
    * `Publish`_
        * `Single`_
        * `Multiple`_
    * `Subscribe`_
        * `Simple`_
        * `Using Callback`_
* `Reporting bugs`_
* `More information`_
   

Installation
------------

The latest stable version is available in the Python Package Index (PyPi) and can be installed using

::

    pip install paho-mqtt

Or with ``virtualenv``:

::

    virtualenv paho-mqtt
    source paho-mqtt/bin/activate
    pip install paho-mqtt

To obtain the full code, including examples and tests, you can clone the git repository:

::

    git clone https://github.com/eclipse/paho.mqtt.python


Once you have the code, it can be installed from your repository as well:

::

    cd paho.mqtt.python
    python setup.py install

Usage and API
-------------

Detailed API documentation is available through **pydoc**. Samples are available in the **examples** directory.

The package provides two modules, a full client and a helper for simple publishing.

Getting Started
***************

Here is a very simple example that subscribes to the broker $SYS topic tree and prints out the resulting messages:

::

    import paho.mqtt.client as mqtt
    
    # The callback for when the client receives a CONNACK response from the server.
    def on_connect(client, userdata, flags, rc):
        print("Connected with result code "+str(rc))

        # Subscribing in on_connect() means that if we lose the connection and
        # reconnect then subscriptions will be renewed.
        client.subscribe("$SYS/#")

    # The callback for when a PUBLISH message is received from the server.
    def on_message(client, userdata, msg):
        print(msg.topic+" "+str(msg.payload))

    client = mqtt.Client()
    client.on_connect = on_connect
    client.on_message = on_message

    client.connect("iot.eclipse.org", 1883, 60)

    # Blocking call that processes network traffic, dispatches callbacks and
    # handles reconnecting.
    # Other loop*() functions are available that give a threaded interface and a
    # manual interface.
    client.loop_forever()

Client
******

You can use the client class as an instance, within a class or by subclassing. The general usage flow is as follows:

* Create a client instance
* Connect to a broker using one of the ``connect*()`` functions
* Call one of the ``loop*()`` functions to maintain network traffic flow with the broker
* Use ``subscribe()`` to subscribe to a topic and receive messages
* Use ``publish()`` to publish messages to the broker
* Use ``disconnect()`` to disconnect from the broker

Callbacks will be called to allow the application to process events as necessary. These callbacks are described below.

Constructor / reinitialise
``````````````````````````

Client()
''''''''

::

    Client(client_id="", clean_session=True, userdata=None, protocol=MQTTv311, transport="tcp")

The ``Client()`` constructor takes the following arguments:

client_id
    the unique client id string used when connecting to the broker. If
    ``client_id`` is zero length or ``None``, then one will be randomly
    generated. In this case the ``clean_session`` parameter must be ``True``.

clean_session
    a boolean that determines the client type. If ``True``, the broker will
    remove all information about this client when it disconnects. If ``False``,
    the client is a durable client and subscription information and queued
    messages will be retained when the client disconnects.
    
    Note that a client will never discard its own outgoing messages on
    disconnect. Calling connect() or reconnect() will cause the messages to be
    resent. Use reinitialise() to reset a client to its original state.

userdata
    user defined data of any type that is passed as the ``userdata`` parameter
    to callbacks. It may be updated at a later point with the
    ``user_data_set()`` function.

protocol
    the version of the MQTT protocol to use for this client. Can be either
    ``MQTTv31`` or ``MQTTv311``

transport
    set to "websockets" to send MQTT over WebSockets. Leave at the default of
    "tcp" to use raw TCP.


Constructor Example
...................

::

    import paho.mqtt.client as mqtt
    
    mqttc = mqtt.Client()


reinitialise()
''''''''''''''

::

    reinitialise(client_id="", clean_session=True, userdata=None)
    
The ``reinitialise()`` function resets the client to its starting state as if it had just been created. It takes the same arguments as the ``Client()`` constructor.

Reinitialise Example
....................

::

    mqttc.reinitialise()

Option functions
````````````````

These functions represent options that can be set on the client to modify its behaviour. In the majority of cases this must be done *before* connecting to a broker.

max_inflight_messages_set()
'''''''''''''''''''''''''''

::

    max_inflight_messages_set(self, inflight)
    
Set the maximum number of messages with QoS>0 that can be part way through their network flow at once.

Defaults to 20. Increasing this value will consume more memory but can increase throughput.

max_queued_messages_set()
'''''''''''''''''''''''''

::

    max_queued_messages_set(self, queue_size)

Set the maximum number of outgoing messages with QoS>0 that can be pending in the outgoing message queue.

Defaults to 0. 0 means unlimited. When the queue is full, any further outgoing messages would be dropped.

message_retry_set()   
'''''''''''''''''''

::

    message_retry_set(retry)

Set the time in seconds before a message with QoS>0 is retried, if the broker does not respond.

This is set to 5 seconds by default and should not normally need changing.

tls_set()
'''''''''

::

    tls_set(ca_certs=None, certfile=None, keyfile=None, cert_reqs=ssl.CERT_REQUIRED,
        tls_version=ssl.PROTOCOL_TLSv1, ciphers=None)

Configure network encryption and authentication options. Enables SSL/TLS support.

ca_certs
    a string path to the Certificate Authority certificate files that are to be treated as trusted by this client. If this is the only option given then the client will operate in a similar manner to a web browser. That is to say it will require the broker to have a certificate signed by the Certificate Authorities in ``ca_certs`` and will communicate using TLS v1, but will not attempt any form of authentication. This provides basic network encryption but may not be sufficient depending on how the broker is configured. By default, on Python 2.7.9+ or 3.4+, the default certification authority of the system is used. On older Python version this parameter is mandatory.

certfile, keyfile
    strings pointing to the PEM encoded client certificate and private keys respectively. If these arguments are not ``None`` then they will be used as client information for TLS based authentication. Support for this feature is broker dependent. Note that if either of these files in encrypted and needs a password to decrypt it, Python will ask for the password at the command line. It is not currently possible to define a callback to provide the password.

cert_reqs
    defines the certificate requirements that the client imposes on the broker. By default this is ``ssl.CERT_REQUIRED``, which means that the broker must provide a certificate. See the ssl pydoc for more information on this parameter.

tls_version
    specifies the version of the SSL/TLS protocol to be used. By default TLS v1 is used. Previous versions (all versions beginning with SSL) are possible but not recommended due to possible security problems.

ciphers
    a string specifying which encryption ciphers are allowable for this connection, or ``None`` to use the defaults. See the ssl pydoc for more information.

Must be called before ``connect*()``.

tls_set_context()
'''''''''''''''''

::

    tls_set_context(context)

Configure network encryption and authentication context. Enables SSL/TLS support.

context
    an ssl.SSLContext object. By default, this is given by ``ssl.create_default_context()``, if available (added in Python 3.4).

If you're unsure about using this method, then either use the default context, or use the ``tls_set`` method. See the ssl module documentation section about `security considerations <https://docs.python.org/3/library/ssl.html#ssl-security>`_ for more information.

Must be called before ``connect*()``.

tls_insecure_set()
''''''''''''''''''

::

    tls_insecure_set(value)

Configure verification of the server hostname in the server certificate.

If ``value`` is set to ``True``, it is impossible to guarantee that the host you are connecting to is not impersonating your server. This can be useful in initial server testing, but makes it possible for a malicious third party to impersonate your server through DNS spoofing, for example.

Do not use this function in a real system. Setting value to True means there is no point using encryption.

Must be called before ``connect*()`` and after ``tls_set()`` or ``tls_set_context()``.

enable_logger()
'''''''''''''''

::

    enable_logger(logger)

Enable logging using the standard python logging package (See PEP 282). This may be used at the same time as the ``on_log`` callback method.

If ``logger`` is specified, then that ``logging.Logger`` object will be used, otherwise one will be created automatically.

Paho logging levels are converted to standard ones according to the following mapping:

====================  ===============
Paho                  logging
====================  ===============
``MQTT_LOG_ERR``      ``logging.ERROR``
``MQTT_LOG_WARNING``  ``logging.WARNING``
``MQTT_LOG_NOTICE``   ``logging.INFO`` *(no direct equivalent)*
``MQTT_LOG_INFO``     ``logging.INFO``
``MQTT_LOG_DEBUG``    ``logging.DEBUG``
====================  ===============

disable_logger()
''''''''''''''''

::

    disable_logger()

Disable logging using standard python logging package. This has no effect on the ``on_log`` callback.

username_pw_set()
'''''''''''''''''

::

    username_pw_set(username, password=None)

Set a username and optionally a password for broker authentication. Must be called before ``connect*()``.

user_data_set()
'''''''''''''''

::

    user_data_set(userdata)

Set the private user data that will be passed to callbacks when events are generated. Use this for your own purpose to support your application.

will_set()
''''''''''

::

    will_set(topic, payload=None, qos=0, retain=False)

Set a Will to be sent to the broker. If the client disconnects without calling
``disconnect()``, the broker will publish the message on its behalf.

topic
    the topic that the will message should be published on.

payload
    the message to send as a will. If not given, or set to ``None`` a zero
    length message will be used as the will. Passing an int or float will
    result in the payload being converted to a string representing that number.
    If you wish to send a true int/float, use ``struct.pack()`` to create the
    payload you require.
    
qos
    the quality of service level to use for the will.
    
retain
    if set to ``True``, the will message will be set as the "last known
    good"/retained message for the topic.

Raises a ``ValueError`` if ``qos`` is not 0, 1 or 2, or if ``topic`` is
``None`` or has zero string length.

Connect / reconnect / disconnect
````````````````````````````````

connect()
'''''''''

::

    connect(host, port=1883, keepalive=60, bind_address="")

The ``connect()`` function connects the client to a broker. This is a blocking
function. It takes the following arguments:

host
    the hostname or IP address of the remote broker

port
    the network port of the server host to connect to. Defaults to 1883. Note
    that the default port for MQTT over SSL/TLS is 8883 so if you are using
    ``tls_set()`` or ``tls_set_context()``, the port may need providing manually

keepalive
    maximum period in seconds allowed between communications with the broker.
    If no other messages are being exchanged, this controls the rate at which
    the client will send ping messages to the broker

bind_address
    the IP address of a local network interface to bind this client to,
    assuming multiple interfaces exist

Callback
........

When the client receives a CONNACK message from the broker in response to the
connect it generates an ``on_connect()`` callback.

Connect Example
...............

::

    mqttc.connect("iot.eclipse.org")

connect_async()
'''''''''''''''

::

    connect_async(host, port=1883, keepalive=60, bind_address="")

Use in conjunction with ``loop_start()`` to connect in a non-blocking manner.
The connection will not complete until ``loop_start()`` is called.

Callback (connect)
..................

When the client receives a CONNACK message from the broker in response to the
connect it generates an ``on_connect()`` callback.

connect_srv()
'''''''''''''

::

    connect_srv(domain, keepalive=60, bind_address="")

Connect to a broker using an SRV DNS lookup to obtain the broker address. Takes
the following arguments:

domain
    the DNS domain to search for SRV records. If ``None``, try to determine the
    local domain name.

See ``connect()`` for a description of the ``keepalive`` and ``bind_address``
arguments.

Callback (connect_srv)
......................

When the client receives a CONNACK message from the broker in response to the
connect it generates an ``on_connect()`` callback.

SRV Connect Example
...................

::

    mqttc.connect_srv("eclipse.org")

reconnect()
'''''''''''

::

    reconnect()

Reconnect to a broker using the previously provided details. You must have
called ``connect*()`` before calling this function.

Callback (reconnect)
....................

When the client receives a CONNACK message from the broker in response to the
connect it generates an ``on_connect()`` callback.

disconnect()
''''''''''''

::

    disconnect()

Disconnect from the broker cleanly. Using ``disconnect()`` will not result in a
will message being sent by the broker.

Callback (disconnect)
.....................

When the client has sent the disconnect message it generates an
``on_disconnect()`` callback.

Network loop
````````````

These functions are the driving force behind the client. If they are not
called, incoming network data will not be processed and outgoing network data
may not be sent in a timely fashion. There are four options for managing the
network loop. Three are described here, the fourth in "External event loop
support" below. Do not mix the different loop functions.

loop()
''''''

::

    loop(timeout=1.0, max_packets=1)

Call regularly to process network events. This call waits in ``select()`` until
the network socket is available for reading or writing, if appropriate, then
handles the incoming/outgoing data. This function blocks for up to ``timeout``
seconds. ``timeout`` must not exceed the ``keepalive`` value for the client or
your client will be regularly disconnected by the broker.

The ``max_packets`` argument is obsolete and should be left unset.

Loop Example
............

::

    run = True
    while run:
        mqttc.loop()
        
loop_start() / loop_stop()
''''''''''''''''''''''''''

::

    loop_start()
    loop_stop(force=False)

These functions implement a threaded interface to the network loop. Calling
``loop_start()`` once, before or after ``connect*()``, runs a thread in the
background to call ``loop()`` automatically. This frees up the main thread for
other work that may be blocking. This call also handles reconnecting to the
broker. Call ``loop_stop()`` to stop the background thread. The ``force``
argument is currently ignored.

Loop Start/Stop Example
.......................

::

    mqttc.connect("iot.eclipse.org")
    mqttc.loop_start()
    
    while True:
        temperature = sensor.blocking_read()
        mqttc.publish("paho/temperature", temperature)

loop_forever()
''''''''''''''

::

    loop_forever(timeout=1.0, max_packets=1, retry_first_connection=False)
    
This is a blocking form of the network loop and will not return until the
client calls ``disconnect()``. It automatically handles reconnecting.

Except for the first connection attempt when using connect_async, use
``retry_first_connection=True`` to make it retry the first connection.
Warning: This might lead to situations where the client keeps connecting to an
non existing host without failing.

The ``timeout`` and ``max_packets`` arguments are obsolete and should be left
unset.

Publishing
``````````

Send a message from the client to the broker.

publish()
'''''''''

::

    publish(topic, payload=None, qos=0, retain=False)

This causes a message to be sent to the broker and subsequently from the broker
to any clients subscribing to matching topics. It takes the following
arguments:

topic
    the topic that the message should be published on

payload
    the actual message to send. If not given, or set to ``None`` a zero length
    message will be used. Passing an int or float will result in the payload
    being converted to a string representing that number. If you wish to send a
    true int/float, use ``struct.pack()`` to create the payload you require
    
qos
    the quality of service level to use

retain
    if set to ``True``, the message will be set as the "last known
    good"/retained message for the topic.

Returns a tuple ``(result, mid)``, where result is ``MQTT_ERR_SUCCESS`` to
indicate success or ``MQTT_ERR_NO_CONN`` if the client is not currently
connected. ``mid`` is the message ID for the publish request. The mid value can
be used to track the publish request by checking against the mid argument in
the ``on_publish()`` callback if it is defined.

A ``ValueError`` will be raised if topic is ``None``, has zero length or is
invalid (contains a wildcard), if ``qos`` is not one of 0, 1 or 2, or if the
length of the payload is greater than 268435455 bytes.

Callback (publish)
..................

When the message has been sent to the broker an ``on_publish()`` callback will
be generated.


Subscribe / Unsubscribe
```````````````````````

subscribe()
'''''''''''

::

    subscribe(topic, qos=0)

Subscribe the client to one or more topics.

This function may be called in three different ways:

Simple string and integer
.........................

e.g. ``subscribe("my/topic", 2)``

topic
    a string specifying the subscription topic to subscribe to.
    
qos
    the desired quality of service level for the subscription. Defaults to 0.

String and integer tuple
........................

e.g. ``subscribe(("my/topic", 1))``

topic
    a tuple of ``(topic, qos)``. Both topic and qos must be present in the tuple.

qos
    not used.

List of string and integer tuples
.................................

e.g. ``subscribe([("my/topic", 0), ("another/topic", 2)])``

This allows multiple topic subscriptions in a single SUBSCRIPTION command,
which is more efficient than using multiple calls to ``subscribe()``.

topic
    a list of tuple of format ``(topic, qos)``. Both topic and qos must be
    present in all of the tuples.
    
qos
    not used.

The function returns a tuple ``(result, mid)``, where ``result`` is
``MQTT_ERR_SUCCESS`` to indicate success or ``(MQTT_ERR_NO_CONN, None)`` if the

ChangeLog.txt

- **BREAKING** Requires Python 2.7 or 3.4+. Closes #163.
=======
v1.3.0 - 2017-06-20
===================

- **BREAKING** Requires Python 2.7 or 3.4+. Closes #163.
- **BREAKING** Remove support for SSL without SSLContext (Requires Python 2.7.9+ or 3.2+).
  Closes #115.
- **BREAKING** on_connect callback is now always called flags. Previously this
  callback could accepts 3 OR 4 arguments, now it must accepts 4. Closes #197.
- **BREAKING** tls_insecure_set() must now be called *after* tls_set()
- Allow username and password to be zero length (as opposed to not being
  present). Closes #80.
- Allow zero length client ids when using MQTT v3.1.1.
- Add SSLContext support, including SNI. Closes #11.
- Improved support for unicode topic and binary payload.
  Closes #15, #16.
- Allow arbitrary Websocket headers and path.
  Closes #169.
- Fix issue with large inbound payload over Websocket. Closes #107.
- Add exponential delay for reconnection. Closes #195.
- Move unit tests to pytest (#164) and tox (#187)
- Add support for standard Python logging. Closes #95.
- Fix duplicate incoming QoS==2 message. Closes #194.

v1.2.3 - 2017-04-20
===================

- Fix possible hang of TLS connection during handshake.
  Closes #3.
- Fix issue with publish helper with TLS connection.
  Closes #180.
- Fix installation issue on non-UTF-8 system.
  Closes #181.
- Fix non-working Websocket over TLS connection.
  Closes #182.
- Move unit tests to pytest (#164) and tox (#187)
=======

v1.2.2 - 2017-04-11
===================

- Fix message lost when using paho.mqtt.publish helper with QoS=0 message.
  Closes #172.

v1.2.1 - 2017-04-03
===================

- Handle unicode username and passwords correctly. Closes #79.
- Fix handling of invalid UTF-8 topics on incoming messages - the library now
  does not attempt to decode the topic - this will happen when the user
  accesses msg.topic in the on_message callback. If the topic is not valid
  UTF-8, an exception will be raised. Closes #75.
- Fix issue with WebSocket connection in case of network issue (timeout or
  connection broken). Closes #105.
- Fix issue with SSL connection, where latest incoming message may be delayed
  or never processed. Closes #131.
- Fix possible message lost with publish.single and publish.multiple. Closes
  #119.


v1.2 - 2016-06-03
=================

- Client.publish() now returns an MQTTMessageInfo object. The MQTTMessageInfo
  object behaves like a tuple of (rc, mid) for backwards compatibility but
  also provides two functions: is_published() and wait_for_published().
  This allows a client to determine whether any given message has been
  published without need for a callback, and also allows the client to block
  waiting until the message has been sent.
- Further fix for Client constructor for the case where "localhost" is
  unresolvable.
- Add paho.mqtt.subscribe module, with simple() and callback() helper
  functions.
- Allow ^C to interrupt client loop.
- Fix for keepalive=0 causing an infinite disconnect/reconnect loop. Closes
  #42.
- Modify callbacks definition/structure to allow classical inheritence. Closes
  #53, #54.
- Add websockets support.
- Default MQTT version is again changed to v3.1.1.
- Client.subscribe() now accepts unicode type topic inputs on Python 2. Closes
  #16.
- paho.mqtt.publish() now raises an MQTTException on a CONNECT failure, rather
  than blindly continuing. Closes #6.
- Don't block on TLS sockets on Python 3. Closes #2.
- Client.publish() now accepts bytes() payloads on Python 3. Closes #1.
- Don't attempt to join() own thread. Closes #14.
- Allow the use of Client.message_callback_add() from inside callbacks. Closes
  #12.
- Use a monotonic time source for keeping track of time, if available. Closes
  #56.


v1.1 - 2015-01-30
=================

- Add support for wildcard certificates. Closes #440547.
- Default connection behaviour has been reverted to MQTT v3.1 instead of
  v3.1.1. There is as yet insufficient support for v3.1.1 to rely on, and
  current v3.1 implementations do not return the correct CONNACK code to allow
  detection of the fault. Closes #451735.
- Fix incorrect handling of queued messages after reconnecting. Closes
  #452672.
- Fix possible race condition if the connection in loop_start() does not
  complete before loop_stop() is called, meaning the network thread never
  ends. Closes #448428. Thanks to Kees Bakker.


v1.0.2 - 2014-09-13
===================

- Fix "protocol" not being used in publish.single()
- Fix Client constructor for the case where "localhost" is unresolvable.
  Closes #439277.
- Don't attempt to encode topic to utf-8 twice. Thanks to Luc Milland.
- Handle "unicode" type payloads on Python 2.7. Thanks to Luc Milland.
- Fix reconnecting after sending more QoS>0 messages than inflight messages is
  set to, whilst connecting.  Closes #443935. Thanks to Hiram van Paassen.
- Fix possible race condition when connecting with TLS and publishing at the
  same time, which could lead to PUBLISH data being sent before any other
  messages and unencrypted. Closes #443964. Thanks to Hiram van Paassen.
- Handle exceptions from select() in client loop() function. Closes #443881.
  Thanks to Jeff Jasper.


v1.0.1
======

- Fix incorrect reconnect that occurred if calling loop_stop() before
  disconnect().

v1.0
====

- Default protocol is now MQTT v3.1.1.
- Client will reconnect using MQTT v3.1 if a v3.1.1 connection fails due to
  the incorrect protocol version number.
- All strings are now correctly encoded as utf-8.
- Add support for "session present" flag in on_connect callback.

v0.9.1
======

- Fix CONNECT packet for MQTT v3.1.1.
- Fix potential lockup when publishing from on_publish callback.
- Add version information to paho.mqtt.__version__


src/paho/mqtt/subscribe.py
# Copyright (c) 2016 Roger Light <roger@atchoo.org>
#
# All rights reserved. This program and the accompanying materials
# are made available under the terms of the Eclipse Public License v1.0
# and Eclipse Distribution License v1.0 which accompany this distribution.
#
# The Eclipse Public License is available at
#    http://www.eclipse.org/legal/epl-v10.html
# and the Eclipse Distribution License is available at
#   http://www.eclipse.org/org/documents/edl-v10.php.
#
# Contributors:
#    Roger Light - initial API and implementation

"""
This module provides some helper functions to allow straightforward subscribing
to topics and retrieving messages. The two functions are simple(), which
returns one or messages matching a set of topics, and callback() which allows
you to pass a callback for processing of messages.
"""

import paho.mqtt.client as paho
import paho.mqtt as mqtt


def _on_connect(client, userdata, flags, rc):
    # pylint: disable=unused-argument
    """Internal callback"""
    if rc != 0:
        raise mqtt.MQTTException(paho.connack_string(rc))

    if isinstance(userdata['topics'], list):
        for topic in userdata['topics']:
            client.subscribe(topic, userdata['qos'])
    else:
        client.subscribe(userdata['topics'], userdata['qos'])


def _on_message_callback(client, userdata, message):
    """Internal callback"""
    userdata['callback'](client, userdata['userdata'], message)


def _on_message_simple(client, userdata, message):
    """Internal callback"""

    if userdata['msg_count'] == 0:
        return

    # Don't process stale retained messages if 'retained' was false
    if message.retain and not userdata['retained']:
        return

    userdata['msg_count'] = userdata['msg_count'] - 1

    if userdata['messages'] is None and userdata['msg_count'] == 0:
        userdata['messages'] = message
        client.disconnect()
        return

    userdata['messages'].append(message)
    if userdata['msg_count'] == 0:
        client.disconnect()


def callback(callback, topics, qos=0, userdata=None, hostname="localhost",
             port=1883, client_id="", keepalive=60, will=None, auth=None,
             tls=None, protocol=paho.MQTTv311, transport="tcp"):
    """Subscribe to a list of topics and process them in a callback function.

    This function creates an MQTT client, connects to a broker and subscribes
    to a list of topics. Incoming messages are processed by the user provided
    callback.  This is a blocking function and will never return.

    callback_in : function of the form "on_message(client, userdata, message)" for
               processing the messages received.

    topics : either a string containing a single topic to subscribe to, or a
             list of topics to subscribe to.

    qos : the qos to use when subscribing. This is applied to all topics.

    userdata : passed to the callback

    hostname : a string containing the address of the broker to connect to.
               Defaults to localhost.

    port : the port to connect to the broker on. Defaults to 1883.

    client_id : the MQTT client id to use. If "" or None, the Paho library will
                generate a client id automatically.

    keepalive : the keepalive timeout value for the client. Defaults to 60
                seconds.

    will : a dict containing will parameters for the client: will = {'topic':
           "<topic>", 'payload':"<payload">, 'qos':<qos>, 'retain':<retain>}.
           Topic is required, all other parameters are optional and will
           default to None, 0 and False respectively.
           Defaults to None, which indicates no will should be used.

    auth : a dict containing authentication parameters for the client:
           auth = {'username':"<username>", 'password':"<password>"}
           Username is required, password is optional and will default to None
           if not provided.
           Defaults to None, which indicates no authentication is to be used.

    tls : a dict containing TLS configuration parameters for the client:
          dict = {'ca_certs':"<ca_certs>", 'certfile':"<certfile>",
          'keyfile':"<keyfile>", 'tls_version':"<tls_version>",
          'ciphers':"<ciphers">}
          ca_certs is required, all other parameters are optional and will
          default to None if not provided, which results in the client using
          the default behaviour - see the paho.mqtt.client documentation.
          Alternatively, tls input can be an SSLContext object, which will be
          processed using the tls_set_context method.
          Defaults to None, which indicates that TLS should not be used.

    transport : set to "tcp" to use the default setting of transport which is
          raw TCP. Set to "websockets" to use WebSockets as the transport.
    """

    if qos < 0 or qos > 2:
        raise ValueError('qos must be in the range 0-2')

    callback_userdata = {
        'callback':callback,
        'topics':topics,
        'qos':qos,
        'userdata':userdata}

    client = paho.Client(client_id=client_id, userdata=callback_userdata,
                         protocol=protocol, transport=transport)
    client.on_message = _on_message_callback
    client.on_connect = _on_connect

    if auth:
        username = auth.get('username')
        if username:
            password = auth.get('password')
            client.username_pw_set(username, password)
        else:
            raise KeyError("The 'username' key was not found, this is "
                           "required for auth")

    if will is not None:
        client.will_set(**will)

    if tls is not None:
        if isinstance(tls, dict):
            client.tls_set(**tls)
        else:
            # Assume input is SSLContext object
            client.tls_set_context(tls)

    client.connect(hostname, port, keepalive)
    client.loop_forever()


def simple(topics, qos=0, msg_count=1, retained=True, hostname="localhost",
           port=1883, client_id="", keepalive=60, will=None, auth=None,
           tls=None, protocol=paho.MQTTv311, transport="tcp"):
    """Subscribe to a list of topics and return msg_count messages.

    This function creates an MQTT client, connects to a broker and subscribes
    to a list of topics. Once "msg_count" messages have been received, it
    disconnects cleanly from the broker and returns the messages.

    topics : either a string containing a single topic to subscribe to, or a
             list of topics to subscribe to.

    qos : the qos to use when subscribing. This is applied to all topics.

    msg_count : the number of messages to retrieve from the broker.
                if msg_count == 1 then a single MQTTMessage will be returned.
                if msg_count > 1 then a list of MQTTMessages will be returned.

    retained : If set to True, retained messages will be processed the same as
               non-retained messages. If set to False, retained messages will
               be ignored. This means that with retained=False and msg_count=1,
               the function will return the first message received that does
               not have the retained flag set.

    hostname : a string containing the address of the broker to connect to.
               Defaults to localhost.

    port : the port to connect to the broker on. Defaults to 1883.

    client_id : the MQTT client id to use. If "" or None, the Paho library will
                generate a client id automatically.

    keepalive : the keepalive timeout value for the client. Defaults to 60
                seconds.

    will : a dict containing will parameters for the client: will = {'topic':
           "<topic>", 'payload':"<payload">, 'qos':<qos>, 'retain':<retain>}.
           Topic is required, all other parameters are optional and will
           default to None, 0 and False respectively.
           Defaults to None, which indicates no will should be used.

    auth : a dict containing authentication parameters for the client:
           auth = {'username':"<username>", 'password':"<password>"}
           Username is required, password is optional and will default to None
           if not provided.
           Defaults to None, which indicates no authentication is to be used.

    tls : a dict containing TLS configuration parameters for the client:
          dict = {'ca_certs':"<ca_certs>", 'certfile':"<certfile>",
          'keyfile':"<keyfile>", 'tls_version':"<tls_version>",
          'ciphers':"<ciphers">}
          ca_certs is required, all other parameters are optional and will
          default to None if not provided, which results in the client using
          the default behaviour - see the paho.mqtt.client documentation.
          Alternatively, tls input can be an SSLContext object, which will be
          processed using the tls_set_context method.
          Defaults to None, which indicates that TLS should not be used.

    transport : set to "tcp" to use the default setting of transport which is
          raw TCP. Set to "websockets" to use WebSockets as the transport.
    """

    if msg_count < 1:
        raise ValueError('msg_count must be > 0')

    # Set ourselves up to return a single message if msg_count == 1, or a list
    # if > 1.
    if msg_count == 1:
        messages = None
    else:
        messages = []

    userdata = {'retained':retained, 'msg_count':msg_count, 'messages':messages}

    callback(_on_message_simple, topics, qos, userdata, hostname, port,
             client_id, keepalive, will, auth, tls, protocol, transport)

    return userdata['messages']
